platforms/software/testing-base/src/main/java/org/gradle/api/tasks/testing/TestEventGenerator.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks.testing;
+
+import org.gradle.api.Incubating;
+
+import java.time.Instant;
+
+/**
+ * Generates test events.
+ *
+ * @since 8.12
+ */
+@Incubating
+public interface TestEventGenerator extends AutoCloseable {
+    /**
+     * Emit a start event for the test.
+     *
+     * @param startTime the time the test started
+     * @since 8.12
+     */
+    void started(Instant startTime);
+
+    /**
+     * Emit a output event for the test. May be emitted multiple times.
+     *
+     * @param destination the destination of the output
+     * @param output some output from the test
+     * @since 8.12
+     */
+    void output(TestOutputEvent.Destination destination, String output);
+
+    /**
+     * Emit a failure event for the test. Does not imply the test has completed.
+     *
+     * @param failure the failure
+     * @since 8.12
+     */
+    void failure(TestFailure failure);
```

Comment:
It looks like we need to rework `TestFailure` a little.  

I'm not sure we want static methods on `TestFailure` to be the thing that creates instances of `TestFailure`. 

We could also consider a `failure(<components of TestFailure>)` and control the creation of `TestFailure` ourselves.

Reply:
Yes, I agree.